cluster/juju/layers/kubernetes-master/actions/create-rbd-pv

```diff
@@ -0,0 +1,207 @@
+#!/usr/bin/python3
+
+from charms.templating.jinja2 import render
+from charms.reactive import is_state
+from charmhelpers.core.hookenv import action_get
+from charmhelpers.core.hookenv import action_set
+from charmhelpers.core.hookenv import action_fail
+from subprocess import check_call
+from subprocess import check_output
+from subprocess import CalledProcessError
+from tempfile import TemporaryDirectory
+import re
+import os
+
+
+def main():
+    ''' Control logic to enlist Ceph RBD volumes as PersistentVolumes in
+    Kubernetes. This will invoke the validation steps, and only execute if
+    this script thinks the environment is 'sane' enough to provision volumes.
+    '''
+
+    # validate relationship pre-reqs before additional steps can be taken
+    if not validate_relation():
+        action_fail('Failed ceph relationship check')
+        return
+
+    context = {}
+
+    context['RBD_NAME'] = action_get_or_default('name').strip()
+    context['RBD_SIZE'] = action_get_or_default('size')
+    context['RBD_FS'] = action_get_or_default('filesystem').strip()
+    context['PV_MODE'] = action_get_or_default('mode').strip()
+
+    # Ensure our paramters match
+    param_validation = validate_parameters(context['RBD_NAME'],
+                                           context['RBD_FS'],
+                                           context['PV_MODE'])
+    if not param_validation == 0:
+        return
+
+    if not validate_unique_volume_name(context['RBD_NAME']):
+        action_fail('Volume name collision detected. Volume creation Aborted.')
+        return
+
+    context['monitors'] = get_monitors()
+
+    # Invoke creation and format the mount device
+    create_rbd_volume(context['RBD_NAME'],
+                      context['RBD_SIZE'],
+                      context['RBD_FS'])
+
+    # Create a temporary workspace to render our persistentVolume template, and
+    # enlist the RDB based PV we've just created
+    with TemporaryDirectory() as active_working_path:
+        temp_template = '{}/pv.yaml'.format(active_working_path)
+        render('rbd-persistent-volume.yaml', temp_template, context)
+
+        check_call(['kubectl', 'create', '-f', temp_template])
+
+
+def action_get_or_default(key):
+    ''' Convenience method to manage defaults since actions dont appear to
+    properly support defaults '''
+
+    value = action_get(key)
+    if value:
+        return value
+    elif key == 'filesystem':
+        return 'xfs'
+    elif key == 'size':
+        return 0
+    elif key == 'mode':
+        return "ReadWriteOnce"
+    else:
+        return ''
+
+
+def create_rbd_volume(name, size, filesystem):
```

Comment:
Replace the CLI calls to rbd with LibRBD

http://docs.ceph.com/docs/jewel/rbd/librbdpy/